File: src/include/asset/ScrapedAsset.php
<?php

namespace asset;

use creator\Creator;
use DateTime;

/**
 * The main asset class.
 * It represents one PBR material, 3D model or other asset.
 */
class ScrapedAsset extends Asset
{
	/**
	 * 
	 * @param null|int $id 
	 * @param null|string $creatorGivenId 
	 * @param string $title 
	 * @param string $url 
	 * @param AssetType $type 
	 * @param Creator $creator 
	 * @param ScrapedAssetStatus $status 
	 * @param array<string> $tags 
	 * @param null|string $rawThumbnailData 
	 * @return void 
	 */
	public function __construct(
		?int $id,
		?string $creatorGivenId,
		string $title,
		string $url,
		AssetType $type,
		Creator $creator,
		public ScrapedAssetStatus $status,
		array $tags = [],
		public ?string $rawThumbnailData = NULL,
	) {
		parent::__construct(
			id: $id,
			creatorGivenId: $creatorGivenId,
			title: $title,
			url: $url,
			type: $type,
			creator: $creator,
			tags: $tags,
		);
	}

	public function toStoredAsset(): StoredAsset
	{
		return new StoredAsset(
			id: $this->id,
			creatorGivenId: $this->creatorGivenId,
			title: $this->title,
			url: $this->url,
			date: new DateTime(),
			type: $this->type,
			creator: $this->creator,
			tags: $this->tags,
			status: $this->status->toStoredAssetStatus(),
			lastSuccessfulValidation: NULL,
		);
	}
}

File: app/asset/scraped_asset.py
from dataclasses import dataclass, field
from typing import Optional, TYPE_CHECKING

from app.asset.asset import Asset
from app.asset.scraped_asset_status import ScrapedAssetStatus

if TYPE_CHECKING:
    from PIL import Image
    from app.asset.stored_asset import StoredAsset


@dataclass
class ScrapedAsset(Asset):
    status: ScrapedAssetStatus = ScrapedAssetStatus.NEWLY_FOUND
    raw_thumbnail: Optional["Image.Image"] = field(default=None, repr=False)

    def to_stored_asset(self) -> "StoredAsset":
        from app.asset.stored_asset import StoredAsset
        from datetime import datetime, timezone

        return StoredAsset(
            id=self.id,
            creator_given_id=self.creator_given_id,
            title=self.title,
            url=self.url,
            date=datetime.now(timezone.utc),
            type=self.type,
            creator=self.creator,
            tags=list(self.tags),
            status=self.status.to_stored_asset_status(),
            last_successful_validation=None,
        )
